providers/default/defltprov.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <openssl/core.h>
+#include <openssl/core_numbers.h>
+#include <openssl/core_names.h>
+
+/* Functions provided by the core */
+static OSSL_core_get_param_types_fn *c_get_param_types = NULL;
+static OSSL_core_get_params_fn *c_get_params = NULL;
+
+/* Parameters we provide to the core */
+static const OSSL_ITEM deflt_param_types[] = {
+    { OSSL_PARAM_UTF8_STRING, OSSL_PROV_PARAM_NAME },
+    { OSSL_PARAM_UTF8_STRING, OSSL_PROV_PARAM_VERSION },
+    { OSSL_PARAM_UTF8_STRING, OSSL_PROV_PARAM_BUILDINFO },
+    { 0, NULL }
+};
+
+static const OSSL_ITEM *deflt_get_param_types(const OSSL_PROVIDER *prov)
+{
+    return deflt_param_types;
+}
+
+static int deflt_get_params(const OSSL_PROVIDER *prov, OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p = params;
+    int ok = 1;
+    static const char provider_name[] = "OpenSSL Default Provider";
+
+    for (; ok && p->key != NULL; p++) {
+        if (strcmp(p->key, OSSL_PROV_PARAM_NAME) == 0) {
```

Comment:
The _params.h_ helpers?

```
return OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(params, OSSL_PROV_PARAM_NAME), provider_name))
    && OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(params, OSSL_PROV_PARAM_VERSION), OPENSSL_VERSION_STR))
    && OSSL_PARAM_set_utf8_string(OSSL_PARAM_locate(params, OSSL_PROV_PARAM_BUILDINFO), OPENSSL_FULL_VERSION_STR));
```

These do buffer size checking, return size setting etc.

Reply:
I think the combined search and get/set helpers have a place still.
This kind of sequence is going to appear a lot thought out the code.

Reply:
New locating helpers in #8519

Reply:
Your use of the params.h functions changes the semantics. What happens if the other end only asks for `OSSL_PROV_PARAM_VERSION`?

To preserve semantics, it should be something like this, no? 

``` C
    const OSSL_PARAM *p;

    return ((p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_NAME)) == NULL || OSSL_PARAM_set_utf8_string(p, provider_name)) 
        && ((p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_VERSION)) == NULL || OSSL_PARAM_set_utf8_string(p, OPENSSL_VERSION_STR)) 
        && ((p = OSSL_PARAM_locate(params, OSSL_PROV_PARAM_BUILDINFO)) == NULL || OSSL_PARAM_set_utf8_string(p, OPENSSL_FULL_VERSION_STR)) 
```

Reply:
My bad.  I forgot a NULL param returned error.  Maybe it should return success in this case.

With the new functions, this would be:

```
return OSSL_PARAM_locate_set_utf8_string(params, OSSL_PROV_PARAM_NAME, provider_name))
    && OSSL_PARAM_locate_set_utf8_string(params, OSSL_PROV_PARAM_VERSION, OPENSSL_VERSION_STR))
    && OSSL_PARAM_locate_set_utf8_string(params, OSSL_PROV_PARAM_BUILDINFO, OPENSSL_FULL_VERSION_STR));
```

That the original missed the length checks, type checks and the return size is a concern about the complexity involved :(

Reply:
Updated based on @paulidale's latest suggestion